File: C++/1083-list_grades.cc
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>


struct Student {
    std::string name, ID;
    int grade;    
};


int lower_bound(const std::vector<Student> &students, int grade) {
    int low = 0, high = students.size() - 1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (students[mid].grade > grade) {
            low = mid + 1;
        } else if (students[mid].grade < grade) {
            high = mid - 1;
        } else {
            return mid;
        }
    }
    return low;
}


int upper_bound(const std::vector<Student> &students, int grade) {
    int low = 0, high = students.size() - 1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (students[mid].grade >= grade) {
            low = mid + 1;
        } else if (students[mid].grade < grade) {
            high = mid - 1;
        }
    }
    return low;
}


int main() {

    std::ios::sync_with_stdio(false);

    int n_student;
    std::cin >> n_student;
    std::vector<Student> students;
    while (n_student--) {
        std::string name, ID;
        int grade;    
        std::cin >> name >> ID >> grade;
        students.push_back({name, ID, grade});
    }
    int bgn_grade, end_grade;
    std::cin >> bgn_grade >> end_grade;
    std::sort(students.begin(), students.end(),
              [](const Student &s1, const Student &s2)
                {return s1.grade > s2.grade;});
    int lower = lower_bound(students, end_grade);
    int upper = upper_bound(students, bgn_grade);
    if (lower != upper) {
        while (lower != upper) {
            std::cout << students[lower].name << ' '
                    << students[lower].ID << '\n';
            ++lower;
        }
    } else {
        std::cout << "NONE\n";
    }

    return 0;
}

File: Python/1083-list_grades.py
from operator import attrgetter


class Student():
    def __init__(self, name, ID, grade):
        self.name = name
        self.ID = ID
        self.grade = grade

    def __str__(self):
        return f'{self.name} {self.ID}'


def students_in_given_grade_range(students, bgn_grade, end_grade):
    def lower_bound(grade):
        low, high = 0, len(students) - 1
        while low <= high:
            mid = (low + high) // 2
            if students[mid].grade > grade:
                low = mid + 1
            elif students[mid].grade < grade:
                high = mid - 1
            else:
                return mid
        return low

    def upper_bound(grade):
        low, high = 0, len(students) - 1
        while low <= high:
            mid = (low + high) // 2
            if students[mid].grade >= grade:
                low = mid + 1
            elif students[mid].grade < grade:
                high = mid - 1
        return low
    students = sorted(students, key=attrgetter('grade'), reverse=True)
    lower = lower_bound(end_grade)
    upper = upper_bound(bgn_grade)
    return students[lower:upper]


if __name__ == '__main__':
    n_student = int(input())
    students = []
    for _ in range(n_student):
        name, ID, grade = input().split()
        students.append(Student(name, ID, int(grade)))
    bgn_grade, end_grade = [int(i) for i in input().split()]
    res = students_in_given_grade_range(students, bgn_grade, end_grade)
    if res:
        for student in res:
            print(student)
    else:
        print("NONE")
